### PR TITLE
Improve XML error handling + Add success message

### PR DESCRIPTION
**Pull request by prckent**
## Proposed changes

Quality of life improvements: To ensure an actual error code is returned for bad XML, replace a few historical error calls with APP_ABORT.  Also print "QMCPACK execution completed successfully" just prior to application return (if no errors/exceptions etc.). Check for this in ctests to catch all early exits.

Motivation for making these changes included some bad XML inputs that I created that did not result in ctest failures.

## What type(s) of changes does this code introduce?

- Bugfix
- New feature
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

sulfur, nightly gcc10 configuration.

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- NA. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCApp/QMCAppBase.cpp [hunk 1]
```diff
@@ -38,6 +38,7 @@ bool QMCAppBase::pushDocument(const std::string& infile)
   else
   {
     app_error() << "File " << infile << " is invalid" << std::endl;
+    APP_ABORT("QMCAppBase::pushDocument");
```
Comment (ye-luo) [inline src/QMCApp/QMCAppBase.cpp:41]: returning false is self-consistent. abort is handled outside. I could not agree with this change.

File: src/QMCApp/QMCMain.cpp [hunk 2]
```diff
@@ -209,7 +209,7 @@ bool QMCMain::execute()
   bool success = validateXML();
   if (!success)
   {
-    ERRORMSG("Input document does not contain valid objects")
+    APP_ABORT("QMCMain::execute. Input document does not contain valid objects");
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:212]: Use myComm->barrier_and_abort()

File: src/QMCApp/QMCMain.cpp [hunk 3]
```diff
@@ -450,12 +450,14 @@ bool QMCMain::validateXML()
           popDocument();
         }
         else
-          myComm->abort();
+        {
+          APP_ABORT("QMCMain::validateXML");
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:454]: Use myComm->barrier_and_abort().

File: src/QMCApp/QMCMain.cpp [hunk 4]
```diff
@@ -450,12 +450,14 @@ bool QMCMain::validateXML()
...
+        }
       }
       else
       {
         app_error() << "tag \"include\" must include an \"href\" attribute." << std::endl;
-        myComm->abort();
+        APP_ABORT("QMCMain::validateXML")
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:460]: Use myComm->barrier_and_abort() including the message sent to app_error.

File: src/QMCApp/QMCMain.cpp [hunk 5]
```diff
@@ -485,17 +487,17 @@ bool QMCMain::validateXML()
...
   if (psiPool->empty())
   {
-    ERRORMSG("Illegal input. Missing wavefunction. ")
+    APP_ABORT("QMCMain::validateXML. Illegal input. Missing wavefunction.");  
     return false;
   }
   if (hamPool->empty())
   {
-    ERRORMSG("Illegal input. Missing hamiltonian. ")
+    APP_ABORT("QMCMain::validateXML. Illegal input. Missing Hamiltonian.");
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:500]: Please use Use myComm->barrier_and_abort().

File: src/QMCApp/qmcapp.cpp [hunk 6]
```diff
@@ -247,6 +247,7 @@ int main(int argc, char** argv)
   }
 
   OHMMS::Controller->finalize();
+  std::cout << "QMCPACK execution completed successfully" << std::endl;
```
Comment (ye-luo) [inline src/QMCApp/qmcapp.cpp:250]: We cannot print on very rank.
Already out to the MPI scope. Very hard to have a consistent logic to send clean output.
Considering we'd better not using this for ctest check. Removal is better for me.

**Review by ye-luo — CHANGES_REQUESTED** [review]
Could you point to or provide a reproducer? Not clear where the issue is precisely.

File: CMake/macros.cmake [hunk 7]
```diff
@@ -99,6 +99,7 @@ FUNCTION( RUN_QMC_APP_NO_COPY TESTNAME WORKDIR PROCS THREADS TEST_ADDED TEST_LAB
         ELSE()
             ADD_TEST( NAME ${TESTNAME} COMMAND ${MPIEXEC_EXECUTABLE} ${MPIEXEC_NUMPROC_FLAG} ${PROCS} ${MPIEXEC_PREFLAGS} ${QMC_APP} ${ARGN} )
             SET_TESTS_PROPERTIES( ${TESTNAME} PROPERTIES FAIL_REGULAR_EXPRESSION "${TEST_FAIL_REGULAR_EXPRESSION}"
+                PASS_REGULAR_EXPRESSION "QMCPACK execution completed successfully"
```
[the remark this answers is not in the document]
Comment (prckent) [reply CMake/macros.cmake:102]: This only tightens the pass criterion (??)

**Comment by prckent** [comment]
Reproducer: Delete the / on a closing XML tag with existing code. e.g. An easy user error such as <random seed="666">

On test criteria, which are only tightened here: see https://cmake.org/cmake/help/latest/prop_test/PASS_REGULAR_EXPRESSION.html and (say ) https://cmake.org/cmake/help/latest/prop_test/SKIP_RETURN_CODE.html . All the pass conditions have to be met and all the fail conditions have to not be met.

Will update the calls and arrange for node 0 only output.

We are going to print something (said in PI voice)

File: CMake/macros.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply CMake/macros.cmake:102]: Tightening is good.

File: src/QMCApp/qmcapp.cpp (hunk 6 above)
Comment (ye-luo) [reply src/QMCApp/qmcapp.cpp:250]: Since this helps tightening the pass criteria, we should add this. Please pay attention to ensemble runs.

**Comment by prckent** [comment]
Maybe fixed. Not sure about ensemble runs, did check conventional MPI runs.

File: src/QMCApp/QMCMain.cpp [hunk 8]
```diff
@@ -209,7 +209,7 @@ bool QMCMain::execute()
   bool success = validateXML();
   if (!success)
   {
-    ERRORMSG("Input document does not contain valid objects")
+    myComm->barrier_and_abort("QMCMain::execute. Input document does not contain valid objects");
     return false;
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:213]: could you remove the return.

File: src/QMCApp/QMCMain.cpp [hunk 9]
```diff
@@ -450,12 +450,13 @@ bool QMCMain::validateXML()
           popDocument();
         }
         else
-          myComm->abort();
+        {
```
Comment (ye-luo) [inline src/QMCApp/QMCMain.cpp:453]: could you remove the {} around a single line if possible.

File: src/QMCApp/qmcapp.cpp [hunk 10]
```diff
@@ -218,8 +218,13 @@ int main(int argc, char** argv)
     else
       validInput = qmc->parse(inputs[0]);
     if (validInput)
+    { 
       qmc->execute();
```
Comment (ye-luo) [inline src/QMCApp/qmcapp.cpp:222]: This execute also returns true/false. Better to check.

**Review by ye-luo — DISMISSED** [review]

**Review by ye-luo — DISMISSED** [review]

**Review by ye-luo — APPROVED** [review]

**Pull request merged**